scripts/ui/controlnet_unit.py

```diff
@@ -0,0 +1,921 @@
+                minimum=0.0,
+                maximum=1.0,
+                interactive=True,
+                elem_id=f"{elem_id_tabname}_{tabname}_controlnet_start_control_step_slider",
+                elem_classes="controlnet_start_control_step_slider",
+            )
+            guidance_end = gr.Slider(
+                label="Ending Control Step",
+                value=self.default_unit.guidance_end,
+                minimum=0.0,
+                maximum=1.0,
+                interactive=True,
+                elem_id=f"{elem_id_tabname}_{tabname}_controlnet_ending_control_step_slider",
+                elem_classes="controlnet_ending_control_step_slider",
+            )
+
+        # advanced options
+        with gr.Column(visible=False) as advanced:
+            processor_res = gr.Slider(
+                label="Preprocessor resolution",
+                value=self.default_unit.processor_res,
+                minimum=64,
+                maximum=2048,
+                visible=False,
+                interactive=False,
+                elem_id=f"{elem_id_tabname}_{tabname}_controlnet_preprocessor_resolution_slider",
+            )
+            threshold_a = gr.Slider(
+                label="Threshold A",
+                value=self.default_unit.threshold_a,
+                minimum=64,
+                maximum=1024,
+                visible=False,
+                interactive=False,
+                elem_id=f"{elem_id_tabname}_{tabname}_controlnet_threshold_A_slider",
+            )
+            threshold_b = gr.Slider(
+                label="Threshold B",
+                value=self.default_unit.threshold_b,
+                minimum=64,
+                maximum=1024,
+                visible=False,
+                interactive=False,
+                elem_id=f"{elem_id_tabname}_{tabname}_controlnet_threshold_B_slider",
+            )
+
+        control_mode = gr.Radio(
+            choices=[e.value for e in external_code.ControlMode],
+            value=self.default_unit.control_mode.value,
+            label="Control Mode",
+            elem_id=f"{elem_id_tabname}_{tabname}_controlnet_control_mode_radio",
+            elem_classes="controlnet_control_mode_radio",
+        )
+
+        resize_mode = gr.Radio(
+            choices=[e.value for e in external_code.ResizeMode],
+            value=self.default_unit.resize_mode.value,
+            label="Resize Mode",
+            elem_id=f"{elem_id_tabname}_{tabname}_controlnet_resize_mode_radio",
+            elem_classes="controlnet_resize_mode_radio",
+        )
+
+        loopback = gr.Checkbox(
+            label="[Loopback] Automatically send generated images to this ControlNet unit",
+            value=self.default_unit.loopback,
+            elem_id=f"{elem_id_tabname}_{tabname}_controlnet_automatically_send_generated_images_checkbox",
+            elem_classes="controlnet_loopback_checkbox",
+        )
+
+        for name, value in locals().items():
```

Comment:
Capture all variable declared and save as class attributes.

Reply:
Did you encounter a problem by just setting the fields we need using `self.enabled = gr...` for example, instead of just `enable = gr...`? It would make the code a little bit easier to read IMO.

Reply:
Sure, I can assign them explicitly.